### PR TITLE
units(dm): allow inject an external storage to units

### PR DESCRIPTION
**Pull request by lance6716**
Signed-off-by: lance6716 <lance6716@gmail.com>

### What problem does this PR solve?

Issue Number: ref #4287

### What is changed and how it works?

as title

### Check List 

Tests 

 - The test will be done at dataflow engine side

Code changes

 - Has exported variable/fields change

### Release note 

```release-note
 `None`.
```

### CODE REVIEW

File: dm/loader/lightning.go [hunk 1]
```diff
@@ -200,7 +200,16 @@ func (l *LightningLoader) runLightning(ctx context.Context, cfg *lcfg.Config) er
...
+	if l.cfg.ExtStorage != nil {
+		err = l.core.RunOnceWithOptions(
+			taskCtx,
+			cfg,
+			lightning.WithDumpFileStorage(l.cfg.ExtStorage),
+			lightning.WithCheckpointStorage(l.cfg.ExtStorage, lightningCheckpointFileName),
+		)
+	} else {
+		err = l.core.RunOnceWithOptions(taskCtx, cfg)
+	}
```
Comment (lichunzhu) [inline dm/loader/lightning.go:212]:
```suggestion
    ops := ...
	if l.cfg.ExtStorage != nil {
		ops = append(ops, 
				lightning.WithDumpFileStorage(l.cfg.ExtStorage),
				lightning.WithCheckpointStorage(l.cfg.ExtStorage, lightningCheckpointFileName))
	}
	err = l.core.RunOnceWithOptions(taskCtx, cfg, ops...)
```

File: dm/dm/config/subtask.go [hunk 2]
```diff
@@ -270,6 +271,9 @@ type SubTaskConfig struct {
 	Experimental struct {
 		AsyncCheckpointFlush bool `yaml:"async-checkpoint-flush" toml:"async-checkpoint-flush" json:"async-checkpoint-flush"`
 	} `yaml:"experimental" toml:"experimental" json:"experimental"`
+
+	// dataflow engine can inject an ExternalStorage to units
+	ExtStorage extstorage.ExternalStorage `toml:"-" json:"-"`
```
Comment (lichunzhu) [inline dm/dm/config/subtask.go:276]: How do we assign this?

**Review by lichunzhu — APPROVED** [review]
LGTM

File: dm/dm/config/subtask.go (hunk 2 above)
Comment (lance6716) [reply dm/dm/config/subtask.go:276]: By public member accessing.

**Review by GMHDBJD — APPROVED** [review]
LGTM

**Pull request merged**